File: LeetCode_Algorithm/1(Two Sum)/1.py
from typing import List

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # sol1
        hash_table = {}
        for idx,num in enumerate(nums):
            match_num = target - num
            if match_num in hash_table:
                return [idx, hash_table[match_num]]     
            hash_table[num] = idx

        # sol2 n^2 시간 제한 
        # for i in range(len(nums)-1):
        #     for j in range(1, len(nums)):
        #         if i!= j and nums[i] + nums[j] == target:
        #             return [i,j]

print(Solution().twoSum([1,2,3], 4))
print(Solution().twoSum([2,7,11,15], 9))

File: LeetCode_Algorithm/1(Two Sum)/1.js


var twoSum = function(nums, target) {
    // javascript는 n^2으로 풀림
    // for (let i=0 ;i < nums.length;i++){
    //     for (let j=1; j< nums.length;j++){
    //         if(i!=j && nums[i]+nums[j] == target){
    //             return [i,j]
    //         }
    //     }
        
    // }
    hashTable = []
    for (let idx = 0; idx < nums.length;idx ++){
        let targetNums = target - nums[idx]
        if (hashTable.includes(targetNums)){
            return [idx, hashTable.indexOf(targetNums)]
        }
        hashTable[idx] = nums[idx]
    }

};

console.log(twoSum([1,2,3], 4))